spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameterManager.scala

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.parameters
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{Callable, Executors, Future, ThreadFactory}
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{Engine, T, Table}
+import org.apache.commons.lang.exception.ExceptionUtils
+import org.apache.log4j.Logger
+import org.apache.spark.sparkExtension.{ParameterManagerMaster, SparkExtension}
+import org.apache.spark.SparkEnv
+import org.apache.spark.storage.{BlockId, BlockManagerWrapper, StorageLevel}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.HashMap
+import scala.reflect._
+
+object AllReduceParameterManager {
+  val logger = Logger.getLogger(getClass)
+
+  private val nextId = new AtomicInteger(0)
+
+  private val pm = new HashMap[String, AllReduceParameterManager]()
+
+  private var executorIdMap: HashMap[String, Int] = null
+
+  def setExecutorMap(map: HashMap[String, Int]): Unit = {
+    executorIdMap = map
+  }
+
+  def get(id: Int, executorId: String): Option[AllReduceParameterManager] = {
+    val eid = id + "exe" + executorIdMap(executorId)
+    pm.get(eid)
+  }
+
+  def createParameterManager[T: ClassTag](executorId: Int, executorNum: Int, partitionNum: Int,
```

Comment:
We actually want to have a single parameter manager running per executor, and have the BigDL program (e.g., the optimizer) to put and get  (potentially multiple) parameters at the PM.